x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { i18n } from '@kbn/i18n';
+
+import { SearchEmbeddable } from 'src/legacy/core_plugins/kibana/public/discover/embeddable/search_embeddable';
+import { ContextMenuAction, ContextMenuActionsRegistryProvider } from 'ui/embeddable';
+import { PanelActionAPI } from 'ui/embeddable/context_menu_actions/types';
+import { kfetch } from 'ui/kfetch';
+import { toastNotifications } from 'ui/notify';
+import { jobCompletionNotifications } from '../lib/job_completion_notifications';
+
+class GetCsvReportPanelAction extends ContextMenuAction {
+  constructor() {
+    super(
+      {
+        displayName: i18n.translate('kbn.dashboard.panel.reportPanel.displayName', {
+          defaultMessage: 'Export to CSV',
+        }),
+        id: 'openReport',
+        parentPanelId: 'mainMenu',
+      },
+      {
+        icon: 'document',
+      }
+    );
+  }
+
+  public async generateJobParams({ searchEmbeddable }: { searchEmbeddable: any }) {
+    // const adapters = searchEmbeddable.getInspectorAdapters();
```

Comment:
There was some discussion on pushing this into the embedabble layer, which I'll do once we get internals ironed-out. Including here just for review-ability sake :)

Reply:
Sorry, I'm not quite sure what this means! I'd like to get my head around it. Can you include some references?

Reply:
I think the basic idea is to have a `generateJobUrl` or some other method inside of `searchEmbeddable`, or really any other embeddable that can have a report generated, instead of these panel action menus doing all of that (for instance this commented code is reaching into private properties and methods inside of the `searchEmbeddable`, so it's breach of concerns in a way).